encoding/migration/unaggregated_iterator.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/m3db/m3metrics/encoding/msgpack"
+	"github.com/m3db/m3metrics/encoding/protobuf"
+)
+
+// UnaggregatedIterator decodes unaggregated metrics.
+type UnaggregatedIterator interface {
+	// Next returns true if there are more items to decode.
+	Next() bool
+
+	// Current returns the current decoded value.
+	Current() encoding.UnaggregatedMessageUnion
+
+	// Err returns the error encountered during decoding, if any.
+	Err() error
+
+	// Close closes the iterator.
+	Close()
+}
+
+type unaggregatedIterator struct {
+	reader encoding.ByteReadScanner
+
+	msgpackIt  msgpack.UnaggregatedIterator
+	protobufIt protobuf.UnaggregatedIterator
+	msg        encoding.UnaggregatedMessageUnion
+	closed     bool
+	err        error
+}
+
+// NewUnaggregatedIterator creates a new unaggregated iterator.
+func NewUnaggregatedIterator(
+	reader encoding.ByteReadScanner,
+	msgpackItOpts msgpack.UnaggregatedIteratorOptions,
+	protobufItOpts protobuf.UnaggregatedOptions,
+) UnaggregatedIterator {
+	msgpackIt := msgpack.NewUnaggregatedIterator(reader, msgpackItOpts)
+	protobufIt := protobuf.NewUnaggregatedIterator(reader, protobufItOpts)
+	return &unaggregatedIterator{
+		reader:     reader,
+		msgpackIt:  msgpackIt,
+		protobufIt: protobufIt,
+	}
+}
+
+func (it *unaggregatedIterator) Next() bool {
+	if it.closed || it.Err() != nil {
+		return false
+	}
+	protocol, err := it.decodeProtocolType()
+	if err != nil {
+		return false
+	}
+	switch protocol {
+	case msgpackProtocol:
+		if !it.msgpackIt.Next() {
+			return false
+		}
+		metric := it.msgpackIt.Metric()
+		policiesList := it.msgpackIt.PoliciesList()
+		msg, err := toUnaggregatedMessageUnion(metric, policiesList)
+		if err != nil {
+			it.err = err
+			return false
+		}
+		it.msg = msg
+		return true
+	case protobufProtocol:
+		if !it.protobufIt.Next() {
+			return false
+		}
+		it.msg = it.protobufIt.Current()
+		return true
+	default:
+		it.err = fmt.Errorf("unexpected protocol type: %v", protocol)
+		return false
+	}
+}
+
+func (it *unaggregatedIterator) Current() encoding.UnaggregatedMessageUnion { return it.msg }
+
+func (it *unaggregatedIterator) Err() error {
+	if it.err != nil {
+		return it.err
+	}
+	if err := it.msgpackIt.Err(); err != nil {
+		return err
+	}
+	return it.protobufIt.Err()
+}
+
+func (it *unaggregatedIterator) Close() {
+	if it.closed {
+		return
+	}
+	it.closed = true
+	it.msg = encoding.UnaggregatedMessageUnion{}
+	it.err = nil
+}
+
+func (it *unaggregatedIterator) decodeProtocolType() (protocolType, error) {
+	// Peek the first byte of the message.
+	b, err := it.reader.ReadByte()
+	if err != nil {
+		it.err = err
+		return unknownProtocolType, err
+	}
+
+	// Put the byte back.
+	if err := it.reader.UnreadByte(); err != nil {
+		it.err = err
+		return unknownProtocolType, err
+	}
+
+	// NB: currently the msgpack encoder always encodes the version first as the first byte
+	// of the msgpack message, and since we are currently at version 1, the first byte will
```

Comment:
So we can't upgrade the version of mspack during the migration right? How long do you envision the full migration from msgpack to just protobuf?